### ISSUE TITLE
eval.data argument in mx.model.FeedForward.create [R-Package]

### ISSUE THREAD

**Issue by zachmayer** [dmlc/mxnet#579]
How do I correctly pass data to `eval.data` in `mx.model.FeedForward.create`?  It seems to take data in a different format than the `X` and `y` arguments, which is confusing:
```{R}
library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  eval.data=list(
    data=test.x,
    label=test.y
  ),
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
```

Yields:
`Error: attempt to apply non-function`

Traceback:
```{R}
> traceback()
2: mx.model.train(symbol, ctx, input.shape, params$arg.params, params$aux.params, 
       1, num.round, optimizer = optimizer, train.data = X, eval.data = eval.data, 
       metric = eval.metric, epoch.end.callback = epoch.end.callback, 
       batch.end.callback = batch.end.callback, kvstore = kvstore)
1: mx.model.FeedForward.create(lro, X = train.x, y = train.y, eval.data = list(data = test.x, 
       data = test.y), ctx = mx.cpu(), num.round = 50, array.batch.size = 20, 
       learning.rate = 2e-06, momentum = 0.9, eval.metric = mx.metric.rmse)
```

**Comment by thirdwing** [comment]
@zachmayer I am afraid this is something we forgot to update.

Maybe the following is what you expect.

```r
eval_data <- mxnet:::mx.model.init.iter(t(train.x), train.y, batch.size=20, is.train = TRUE)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  eval.data=eval_data,
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
```

```
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[1] Train-rmse=18.7544198250762
[1] Validation-rmse=17.6765946079206
[2] Train-rmse=13.5563010119916
[2] Validation-rmse=10.0524487177531
[3] Train-rmse=11.0570303807623
......
```

**Comment by zachmayer** [comment]
ahah!  Thank you.

In general, examples like that are really useful in the documentation.  Thanks!

**Comment by tqchen** [comment]
@thirdwing  I think we can optionally support passing in a list into eval_data

**Comment by thirdwing** [comment]
Yes, adding now. Just a few lines.

**Comment by thirdwing** [comment]
@zachmayer Now you can use your old code.

```r
library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  eval.data=list(
    data=test.x,
    label=test.y
  ),
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
```
```
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[1] Train-rmse=16.0632823518301
[1] Validation-rmse=10.1766422920373
[2] Train-rmse=12.279237506027
[2] Validation-rmse=12.4331790778543
[3] Train-rmse=11.1984636779466
[3] Validation-rmse=10.3303056488224
......
```

**Comment by zachmayer** [comment]
Fantastic, thank you.  Would it be difficult to implement early stopping based on the evaluation set?

**Comment by tqchen** [comment]
@zachmayer MXNetR is a R-package, and all the training logic, including the parameter optimization, evaluation are on R, so I suppose it is possible for any R hacker to add that feature :)

**Comment by zachmayer** [comment]
I'll take a look :-)



—

**Comment by thirdwing** [comment]
@zachmayer Actually it is quite easy to stop early. 

Just add a new callback `mx.callback.early.stop`.

It takes two arguments, `train.metric` and `eval.metric`. The training will stop once one of them goes below the standard you set.

```r
library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  eval.data=list(
    data=test.x,
    label=test.y
  ),
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse,
  epoch.end.callback=mx.callback.early.stop(10, 10))
```
```
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[1] Train-rmse=16.0632823518301
[1] Validation-rmse=10.1766422920373
[2] Train-rmse=12.279237506027
[2] Validation-rmse=12.4331790778543
[3] Train-rmse=11.1984636779466
[3] Validation-rmse=10.3303056488224
[4] Train-rmse=10.2645248908194
[4] Validation-rmse=8.4276043147402
```

**Comment by zachmayer** [comment]
Ok, I'm starting to get how callbacks work.  Can you share the source code for `mx.callback.early.stop`?  I'll probably modify it to stop after N iterations without an improvement in validation error, rather than a hard threshold for validation error.

**Comment by thirdwing** [comment]
It is on my github https://github.com/thirdwing/mxnet/commit/76c6d7bdf59ecae78a7c16612c4775f7427ae8ac

You can see the difference from the link. I am writing a document for callback functions. There are many things can be done through callbacks.
